contents/c++/rvalue.md

```diff
@@ -0,0 +1,145 @@
+in order to determine whether the parameters given are lvalues or rvalues:
+
+```cpp
+void print(int& lvalue) { // takes an lvalue
+    std::cout << "lvalue method used";
+}
+
+void print(int&& rvalue) { // take a rvalue
+    std::cout << "rvalue method used";
+}
+
+int main() {
+    int x = 5;
+    print(x); // lvalue method used
+    print(10); // rvalue method used.
+}
+```
+
+Using of these functions overload with rvalue parameters, which take on temporary objects, helps us to write more efficient program using Move Semantics!
+
+### Move Semantics
+The main usage of rvalue references is that it allows us to create *move* constructor and *move* assignments, instead of *copy* constructor and *copy* assignments by default. Since rvalues are typically temporary objects, we can just *move* the value instead of *copying* it, thus helping us to save on memory consumption and improving on the performance on our program as well!
+
+One example implementation of a move constructor and move assignment is:
+```cpp
+Foo(Foo&& other) { // move constructor
+    x = other.x;
+    y = other.y;
+    z = other.z;
+}
+
+Foo& operator=(Foo&& other) { // move assignment
+    x = std::move(other.x);
+    y = std::move(other.y);
+    z = std::move(other.z);
+    return *this;
+}
+```
+
+In this move constructor and assignment, the contents of the `other` parameter is *moved* into the object, and the contents in `other` is destroyed afterwards. No additional memory allocation occurs, and the *move* operation is done quickly by a few assignment of address operations, leading to a more performance and memory efficient program.
+
+Along with Move Semantics in C++11, the [STL library](https://www.geeksforgeeks.org/the-c-standard-template-library-stl/) functions also defines the overloaded *move* functions for its container classes, thus we can take advantage of these Move Semantics by simply supplying rvalues, without the need to redefine the classes ourselves.
+
+## Rvalue anti-pattern
+After learning about the new rvalue references and Move Semantics in C++11, many programmers from the older C++ eras may fall into a trap of the rvalue [anti-pattern](https://en.wikipedia.org/wiki/Anti-pattern).
+
+Consider these 2 class constructors:
+```cpp
+Foo(const std::string& x, const std::string& y) { // copy constructor
+    _x = x;
+    _y = y;
+}
+
+Foo(std::string&& x, std::string&& y) { // move constructor
+    _x = x;
+    _y = y;
+}
+```
+
+As we can see, the *move* constructor defines each of its parameter to be rvalues. In this case, if we were to call the constructor using a mixture of both, lvalue for `x` and rvalue for `y`, the copy constructor instead will be called! This is because the copy constructor and take both lvalues and rvalues, both the *move* constructor can only take rvalues. As such, we may think we have made use of Move Semantics to optimize our program, but that may not be the case!
```

Comment:
>This is because the copy constructor and take both lvalues and rvalues, both the *move* constructor can only take rvalues.

This sentence is rather hard to understand as well.